src/function/maybe_changed_after.rs

```diff
@@ -73,22 +114,29 @@ where
         );
 
         // Check if the inputs are still valid and we can just compare `changed_at`.
-        if self.deep_verify_memo(db, &old_memo, &active_query) {
-            return Some(old_memo.revisions.changed_at > revision);
+        let active_query = zalsa_local.push_query(database_key_index);
```

Comment:
From looking at the red knot benchmarks, the regression mainly comes from the extra `push_query` calls here (that probably also applies for queries not participating in cycles) and the constructed hash set in `deep_verify_memo` (specific to red knot?)

Reply:
Could we move the `push_query` call into `deep_verify_memo` after the second `shallow_verify_memo` or does that result in deadlocks? Just so that we can avoid pushing queries unless it's absolutely necessary